### ISSUE TITLE
Can I use this package with fast.ai?

### ISSUE THREAD

**Issue by wshuyi** [huggingface/accelerate#59]
Hi! It's so exciting! Great job! May I know if I can use this package with fast.ai? Any Document for that? Thanks! :)

**Comment by sgugger** [comment]
Hi there! It's probably not compatible with fastai as the fastai library uses their own `DataLoader` and `Optimizer` classes (different from the PyTorch ones). We're thinking of how we can best add support for fastai so stay tuned!

**Comment by wshuyi** [comment]
Okay. Thanks for the reply! :)

**Comment by vishalghor** [comment]
okay, Thank you :)

**Comment by muellerzr** [comment]
Support for fastai is included in this PR: https://github.com/fastai/fastai/pull/3646

**Comment by costantinoai** [comment]
Is there any news on this? 

[https://docs.fast.ai/distributed.html](Here) on the fastai docs, they say that it is possible to use accelerate with fastai. I tried it and it works. I am able to train the model faster. The only problem is that the losses go down way more slowly, and the accuracy barely improves when I run from CLI (as opposed to running the same exact script from IDE, that in my understanding is not using accelerate). Is it some sort of bug on fastai side?

**Comment by muellerzr** [comment]
@costantinoai when you say in notebook do you mean using the notebook_launcher? Or training on a single GPU

**Comment by costantinoai** [comment]
Thanks for your prompt reply.

I didn't mention the notebook. I am training on command line with accelerate, and on IDE (Spyder) without accelerate, using the exact same learning rate. From command line it runs faster, but from IDE the losses go down more quickly. I am following the instructions on the link I posted in the previous comment.

**Comment by muellerzr** [comment]
We can't do much without seeing the code you trained with, can you provide a sample? And perhaps use a dataset that's public to try and recreate it? Like one of the distributed app examples in the fastai docs here:

https://docs.fast.ai/distributed_app_examples

**Comment by muellerzr** [comment]
What's the output of `accelerate env` here?

**Comment by costantinoai** [comment]
- `Accelerate` version: 0.10.0
- Platform: Linux-4.15.0-76-generic-x86_64-with-glibc2.27
- Python version: 3.9.12
- Numpy version: 1.22.3
- PyTorch version (GPU?): 1.11.0 (True)
- `Accelerate` default config:
	- compute_environment: LOCAL_MACHINE
	- distributed_type: MULTI_GPU
	- mixed_precision: fp16
	- use_cpu: False
	- num_processes: 3
	- machine_rank: 0
	- num_machines: 1
	- main_process_ip: None
	- main_process_port: None
	- main_training_function: main
	- deepspeed_config: {}
	- fsdp_config: {}

**Comment by costantinoai** [comment]
This is the output from CLI with fp16 set to no:

> epoch     train_loss  valid_loss  accuracy  top_k_accuracy  time    
> 0         2.037301    2.129598    0.251592  0.782786        00:09                                                                                     
> 1         1.847155    1.756253    0.390374  0.887955        00:08                                                                                     
> 2         1.730854    1.466629    0.513369  0.912911        00:08                                                                                     
> 3         1.645347    1.365647    0.571174  0.923860        00:08                                                                                     
> 4         1.569351    1.582687    0.516425  0.931500        00:08                                                                                     
> 5         1.509384    1.225409    0.609626  0.938375        00:08                                                                                     
> 6         1.449147    1.243672    0.612172  0.934810        00:08                                                                                     
> 7         1.398242    1.166874    0.627706  0.949071        00:08                                                                                     
> 8         1.349698    1.050111    0.671760  0.950344        00:08                                                                                     
> 9         1.294484    0.942374    0.712758  0.965623        00:08  

here the output of `accelerate env`

> - `Accelerate` version: 0.10.0
> - Platform: Linux-4.15.0-76-generic-x86_64-with-glibc2.27
> - Python version: 3.9.12
> - Numpy version: 1.22.3
> - PyTorch version (GPU?): 1.11.0 (True)
> - `Accelerate` default config:
> 	- compute_environment: LOCAL_MACHINE
> 	- distributed_type: MULTI_GPU
> 	- mixed_precision: no
> 	- use_cpu: False
> 	- num_processes: 3
> 	- machine_rank: 0
> 	- num_machines: 1
> 	- main_process_ip: None
> 	- main_process_port: None
> 	- main_training_function: main
> 	- deepspeed_config: {}
> 	- fsdp_config: {}

**Comment by costantinoai** [comment]
Sorry if I asked in the wrong issue then. I didn't see any mention of TPU in the first comment (or in any other comment for that matter).

Thanks for the support. I will keep an eye on the open issues.

**Comment by muellerzr** [comment]
Hey @costantinoai, I'm writing a doc on this soon but this is a very common mistake when it comes to benchmarking on a single GPU vs multiple GPUs. On Spyder you're not using accelerate, which means you aren't training in a distributed fashion.

To see equal results, the batch size on a single GPU needs to be equivalent to bs*num_gpus on the multi-gpu system. This means that to test a bs of 64 on two GPUs, the single GPU needs a bs of 128. You can see below the results of adding the following line to your code, and the times and results are almost identical: 

```python
if torch.cuda.device_count() > 1:
    print(f"{torch.cuda.device_count()} devices found, using bs=64")
    bs = 64
else:
    print(f"{torch.cuda.device_count()} devices found, using bs=128")
    bs = 128
```
(And then pass bs to the dataloader constructor)

![image](https://user-images.githubusercontent.com/7831895/193048065-c6ef3a85-3e01-4abe-abf9-0f7c092d11d0.png)

**Comment by costantinoai** [comment]
Yep, I also did some testing at the time and I realized the batch size problem. But thanks for looking into this! Indeed, I guess that more clear information about this common mistake in the documentation would be helpful to others in the future 😄

**Comment by muellerzr** [comment]
Hey @costantinoai , so you don't look insane I'll reiterate what I said, so hopefully others can find it.

When benchmarking, it's *very* important to take into account the batch size. So if I test with a bs of 64 on two GPUs, I should test on a single GPU with a bs of 128, and you will get similar results:

![image](https://user-images.githubusercontent.com/7831895/193050663-dd6f1818-6396-435e-bf70-1cdfe1be241c.png)


I'll be writing a benchmarking doc on all of this soon :)
